src/app/app.component.ts

```diff
@@ -1,10 +1,78 @@
-import { Component } from '@angular/core';
+import { Component, OnInit } from '@angular/core';
+import { FormGroup, FormBuilder, FormControl, FormArray, Validators, ValidatorFn, AbstractControl } from '@angular/forms';
+import { HeroService } from './test-service/hero.service';
+import { Hero, HeroPower } from './test-service/hero.interface';
+import { Utils } from './utils';
 
 @Component({
-  selector: 'app-root',
-  templateUrl: './app.component.html',
-  styleUrls: ['./app.component.css']
+    selector: 'app-root',
+    templateUrl: './app.component.html',
+    styleUrls: ['./app.component.css']
 })
-export class AppComponent {
-  title = 'reactive-forms-example';
+export class AppComponent implements OnInit {
+    loading = true;
+    results = 'No results yet, click on save';
+
+    form: FormGroup;
+
+    static createPowerControl(name: string, description: string): FormGroup {
+        return new FormGroup({
+            name: new FormControl(name, Validators.required),
+            description: new FormControl(description),
+        });
+    }
+
+    constructor (
+        private heroService: HeroService,
+    ) { }
+
+    ngOnInit() {
+        this.heroService.getAsyncData()
+            .subscribe(hero => {
+                this.initForm(hero);
+                this.loading = false;
+            });
+    }
+
+    initForm(hero: Hero) {
+        const powerControls = hero.powers.map(power => AppComponent.createPowerControl(power.name, power.description));
+
+        this.form = new FormGroup({
+            name: new FormGroup({
+                alias: new FormControl(hero.name.alias, Validators.required),
+                realName: new FormControl(hero.name.realName),
+            }),
+            appearances: new FormGroup({
+                comics: new FormControl(hero.appearances.comics),
+                movies: new FormControl(hero.appearances.movies),
+                games: new FormControl(hero.appearances.games),
+            }),
+            powers: new FormArray(powerControls, this.getPowersValidator()),
```

Comment:
Are there no built-in validators for array length? Neither `Validators.min` nor `Validators.minLength` (and the `max` equivalents) validated this restriction correctly.